### PR TITLE
Raise on deprecation, fix Ember Data default serializer deprecation

### LINKED ISSUE #129: Flag for review - Deprecation warning in Ember CLI test environment
I'm at the end of the Ember Data section in part 2 of the tutorial. We've just built our own adapter to handle the JSON data and I got this warning in the terminal after running the `ember t -s` command:

```
DEPRECATION: store.serializerFor("rental") resolved the "-json-api" serializer via the deprecated `adapter.defaultSerializer` property.

	Previously, if no application or type-specific serializer was specified, the store would attempt to lookup a serializer via the `defaultSerializer` property on the type's adapter. This behavior is deprec favor of explicitly defining a type-specific serializer or application serializer [deprecation id: ember-data:default-serializer] See https://deprecations.emberjs.com/ember-data/v3.x#toc_ember-data:default-seri
alizers for more details.
        at logDeprecationStackTrace (http://localhost:7357/assets/vendor.js:38578:21)
        at HANDLERS.<computed> (http://localhost:7357/assets/vendor.js:38675:9)
        at raiseOnDeprecation (http://localhost:7357/assets/vendor.js:38605:9)
        at HANDLERS.<computed> (http://localhost:7357/assets/vendor.js:38675:9)
        at invoke (http://localhost:7357/assets/vendor.js:38687:9)
        at Object.deprecate (http://localhost:7357/assets/vendor.js:38643:28)
        at DefaultStore.serializerFor (http://localhost:7357/assets/vendor.js:87670:37)
        at http://localhost:7357/assets/vendor.js:84377:30
        at invokeCallback (http://localhost:7357/assets/vendor.js:62786:17)DEPRECATION: store.serializerFor("rental") resolved the "-json-api" serializer via the deprecated `adapter.defaultSerializer` property.

        Previously, if no application or type-specific serializer was specified, the store would attempt to lookup a serializer via the `defaultSerializer` property on the type's adapter. This behavior is deprec favor of explicitly defining a type-specific serializer or application serializer [deprecation id: ember-data:default-serializer] See https://deprecations.emberjs.com/ember-data/v3.x#toc_ember-data:default-seri
alizers for more details.
        at logDeprecationStackTrace (http://localhost:7357/assets/vendor.js:38578:21)
        at HANDLERS.<computed> (http://localhost:7357/assets/vendor.js:38675:9)
        at raiseOnDeprecation (http://localhost:7357/assets/vendor.js:38605:9)
```

All my tests are passing as expected, but the tutorial didn't make mention of this deprecation warning so I thought it should be flagged here. I don't have a suggested fix here, just a flag for someone with experience to review.

### PR DESCRIPTION
**Pull request by chancancode**
Fixes #129

### CODE REVIEW

**Comment by chancancode** [comment]
@jafarlow would you be able to review this change?

**Comment by jafarlow** [comment]
@chancancode It looks like you've added in the serializer code with a note on why it's there. That should fix my issue.

**Comment by chancancode** [comment]
@jafarlow Yes, thank you! Just wanted to check if the prose makes sense to you and/or if I made any typos.

**Comment by jafarlow** [comment]
It looks like you removed the "However, as mentioned above..." statement. Should that be removed or should it stay? It looks like it's explaining why we need to create a customized adapter.

Other than that it all looks good to my eyes.

**Comment by chancancode** [comment]
Yes, that was intentional, since the text was written as if we were just creating one file (it was just talking about the adapter). I was going to amend it, but it seems unnecessary on second read. The part where it said "there are some minor differences" is also covered below, imo. But if you think it's less clear now, we can iterate on making it better.

**Comment by jafarlow** [comment]
Just beneath the new code is this phrase, "Typically, this should be sufficient, but as mentioned above, Ember Data also allows us to define model-specific adapters." As I read it, this is matched to the statement you've removed, though I could be wrong.

This one could be amended and placed before the code for both adapter and serializer. Something like, "Ember Data allows us to define model-specific adapters so we can customize the default behavior by defining our own adapter, like we'll do right now." 

Or, we could remove that mention altogether. Whichever you think would be the simplest way for someone new to understand the "why" behind the "what" here.

---

(removed text):

> "However, as mentioned above, there are some minor differences between how our server works and Ember Data's default assumptions. We can customize the default behavior by defining our own adapter:"

(subsequent text I think is directly linked):

> "By convention, adapters are located at `app/adapters`. Furthermore, the adapter named `application` is called the *application adapter*, which will be used to fetch data for all models in our app. Typically, this should be sufficient, but as mentioned above, Ember Data also allows us to define model-specific adapters. For example, if `app/adapters/rental.js` is present, it will be used for fetching the `rental` model. We won't be needing this feature, but it's good to know that it exists."

**Comment by chancancode** [comment]
It's actually saying a slightly different thing, so thank you for pointing out the potential confusion.

There are a couple of things going on there:

1. Ember Data uses the JSON:API adapter/serializer by default, which comes with some default assumptions
2. If you are using JSON:API on the backend, but the assumptions doesn't match up, you can customize the default JSON:API adapter/serializer by subclassing
3. You can do even more fine-grained customization by defining model-specific adapter/serializers that overrides the application adapter/serializer when fetching data for that particular model (e.g. you can have `Rental`s coming from backend A and `User`s coming from a different backend B)

The line you pointed out was trying to interject (3) in the middle of doing (2). Since we don't need that feature (and it's not _that_ common to be needing it), it's really an unnecessary side point. I tried to relocate everything about model-specific adapters/serializers to the very end of the chapter.

Does the latest edits make more sense to you now?


It was trying to interject a side point: application adapter/serializer is what ED uses by default, and you can have more precise controlfurther customize particular

**Comment by jafarlow** [comment]
I just read through the changes. This is awesome and make a lot more sense to me now. Thank you for addressing it so promptly!

File: src/markdown/tutorial/part-2/11-ember-data.md [hunk 1]
```diff
@@ -290,17 +297,16 @@ Adding a namespace prefix happens to be pretty common across Ember apps, so the
...
 
-> Zoey says...
->
-> Since the JSON data returned by our server is JSON:API-compliant, the default serializer (which would be [`JSONAPISerializer`][TODO: link to docs]) should work just fine for us. However, if we had wanted to customize it, we could create an `application` serializer at `app/serializers/application.js`.
+Similarly, serializers are located at `app/serializers`. Adapters and serializers are always added together as a pair. We added an `application` adapter, so we also added a corresponding serializer to go with it as well. Since the JSON data returned by our server is JSON:API-compliant, the default [`JSONAPISerializer`][TODO: link to docs] work just fine for us without further customization.
```
Comment (lifeart) [inline src/markdown/tutorial/part-2/11-ember-data.md]: todo?

**Comment by runspired** [comment]
Overall this seems like an improvement, although I wish it wasn't quite so "sub-class" centric. It might be worth reviewing the api docs overview and minimum interface for adapters and serializers.

**Pull request merged**